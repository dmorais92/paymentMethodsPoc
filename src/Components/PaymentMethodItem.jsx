import React, { useState } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import StyledComponents from './Components.styled';
import { ReactComponent as DownIcon } from './Icons/DownIcon.svg';
import { ReactComponent as DeleteIcon } from './Icons/DeleteIcon.svg';
import InlineEditInput from './InlineEditInput';

const {
  Button,
  ListItem,
  ListItemSubtitle,
  ListItemTitleSubtitle,
  ListItemTitle,
} = StyledComponents;

const StyledDeleteIcon = styled(DeleteIcon)`
  && {
    height: 24px;
    width: 24px;
    border-radius: 50px;
    border: 2px solid #ddd;
    position: relative;
    bottom: 12px;

    > :nth-child(2) {
      fill: #f44336;
    }
    &:hover {
      > :nth-child(2) {
        fill: #c62828;
      }
    }
  }
`;

const PaymentMethodItem = (props) => {
  const [expanded, toggleExpanded] = useState(false);
  const ExpandedSection = styled.div`
    margin: 6px;
    height: auto;
    width: 100%;
    opacity: 1;
    max-height: 250px;
    overflow-y: auto;
    overflow-x: hidden;
    display: flex;
    flex-flow: row wrap;
    justify-content: space-between;
    div {
      margin: 0px 5px 5px 0px;
    }
  `;
  const {
    onChange,
    onDelete,
    amount,
    currency,
    reference,
    date,
    purpose,
    type,
    id,
    paymentType,
    debtor,
    benificiary,
    charges,
  } = props;
  return (
    <ListItem>
      <ListItemTitleSubtitle>
        <ListItemTitle>{amount}</ListItemTitle>
        <ListItemSubtitle>{currency}</ListItemSubtitle>
      </ListItemTitleSubtitle>
      <ListItemTitleSubtitle>
        <ListItemTitle data-test-id="reference-title">{reference}</ListItemTitle>
        <ListItemSubtitle>{date}</ListItemSubtitle>
      </ListItemTitleSubtitle>
      <ListItemTitleSubtitle>
        <Button plain onClick={() => onDelete(id)}>
          <StyledDeleteIcon />
        </Button>
        <Button data-test-id="expand-button" onClick={() => toggleExpanded(!expanded)}>
          <DownIcon className={(expanded && 'inverted') || ''} />
        </Button>
      </ListItemTitleSubtitle>
      {expanded && (
        <ExpandedSection>
          <ListItemTitleSubtitle>
            <ListItemTitle>Transaction</ListItemTitle>
            <ListItemSubtitle>{type}</ListItemSubtitle>
          </ListItemTitleSubtitle>
          <ListItemTitleSubtitle>
            <ListItemTitle>Reference</ListItemTitle>
            <InlineEditInput
              value={reference}
              onChange={(val) => onChange(val, 'reference', id)}
              viewProp="reference"
            />
          </ListItemTitleSubtitle>
          <ListItemTitleSubtitle>
            <ListItemTitle>Purpose</ListItemTitle>
            <InlineEditInput
              value={purpose}
              onChange={(val) => onChange(val, 'purpose', id)}
              viewProp="purpose"
            />
          </ListItemTitleSubtitle>
          <ListItemTitleSubtitle>
            <ListItemTitle>Payment Type</ListItemTitle>
            <ListItemSubtitle>{paymentType}</ListItemSubtitle>
          </ListItemTitleSubtitle>
          <ListItemTitleSubtitle>
            <ListItemTitle>From:</ListItemTitle>
            <ListItemSubtitle>{debtor.account_name}</ListItemSubtitle>
            <ListItemTitle>To:</ListItemTitle>
            <ListItemSubtitle>{benificiary.account_name}</ListItemSubtitle>
          </ListItemTitleSubtitle>
          <ListItemTitleSubtitle>
            <ListItemTitle>Amount</ListItemTitle>

            <ListItemSubtitle>
              {amount}
              {' '}
              {currency}
            </ListItemSubtitle>
            <ListItemTitle>Receiver Charges</ListItemTitle>
            <ListItemSubtitle>
              {`+${charges.receiver_charges_amount} ${charges.receiver_charges_currency}`}
            </ListItemSubtitle>
            {
              charges && charges.sender_charges && charges.sender_charges.length && (
                <>

                  <ListItemTitle>Sender Charges</ListItemTitle>
                  {
                charges.sender_charges.map((charge) => (
                  <ListItemSubtitle>
                    {`+${charge.amount} ${charge.currency}`}
                  </ListItemSubtitle>
                ))
              }
                </>
              )
            }
          </ListItemTitleSubtitle>
        </ExpandedSection>
      )}
    </ListItem>
  );
};

PaymentMethodItem.propTypes = {
  type: PropTypes.string,
  amount: PropTypes.string,
  id: PropTypes.string,
  benificiary: PropTypes.shape({
    account_name: PropTypes.string,
  }),
  charges: PropTypes.shape({
    receiver_charges_amount: PropTypes.string,
    receiver_charges_currency: PropTypes.string,
    sender_charges: PropTypes.arrayOf(
      PropTypes.shape({
        amount: PropTypes.string,
        currency: PropTypes.string,
      }),
    ),
  }),
  debtor: PropTypes.shape({
    account_name: PropTypes.string,
  }),
  currency: PropTypes.string,
  purpose: PropTypes.string,
  paymentType: PropTypes.string,
  date: PropTypes.string,
  reference: PropTypes.string,
  onChange: PropTypes.func,
  onDelete: PropTypes.func,
};

PaymentMethodItem.defaultProps = {
  onChange: () => null,
  onDelete: () => null,
  id: '0',
  currency: '',
  purpose: 'No purpose was given',
  type: 'Payment',
  amount: '0',
  paymentType: 'Credit',
  date: 'Today',
  reference: 'Empty reference',
  benificiary: {
    account_name: '---',
  },
  charges: {
    receiver_charges_amount: '',
    receiver_charges_currency: '',
    sender_charges: [],
  },
  debtor: {
    account_name: '---',
  },
};

export default PaymentMethodItem;
