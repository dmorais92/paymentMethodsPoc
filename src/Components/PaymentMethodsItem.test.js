import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import PaymentMethodItem from './PaymentMethodItem';

const testProps = {
	type: 'Payment',
	currency: '€',
	amount: '435.55',
	id: 'test_id_1234',
	purpose: 'Shopping payment',
	paymentType: 'VISA',
	reference: 'Groceries',
	date: '23-02-2018',
	beneficiary: {
		account_name: 'Test'
	},
	debtor: {
		account_name: 'Mr. Debtor'
	},
	charges: {
		receiver_charges_ammount: '34',
		receiver_charges_currency: '€'
	}
};

describe('PaymentMethodItem component', () => {
	const toggleExpanded = jest.fn();
	const useStateSpy = jest.spyOn(React, 'useState');
	useStateSpy.mockImplementation(init => [init, toggleExpanded]);
	it('renders without crashing', () => {
		const div = document.createElement('div');
		ReactDOM.render(<PaymentMethodItem />, div);
		ReactDOM.unmountComponentAtNode(div);
	});
	it('matches snapshot', () => {
		const component = renderer.create(<PaymentMethodItem />);
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});
	it('renders with test props', () => {
		const div = document.createElement('div');
		ReactDOM.render(<PaymentMethodItem {...testProps} />, div);
		ReactDOM.unmountComponentAtNode(div);
	});
	it('expands and renders text inputs', () => {
		const component = renderer.create(<PaymentMethodItem {...testProps} />);
		const testInstance = component.root;
		const expandButton = testInstance.findByProps({
			'data-test-id': 'expand-button'
		});
		renderer.act(() => {
			expandButton.props.onClick();
		});
		expect(
			testInstance.findAllByProps({ 'data-test-id': 'inline-inputs' }).length
		).toBe(2);
	});
	it('changing reference text input will also change reference title', () => {
		const component = renderer.create(
			<PaymentMethodItem
				{...testProps}
				onChange={val => {
					testProps.reference = val;
				}}
			/>
		);
		const testInstance = component.root;
		const expandButton = testInstance.findByProps({
			'data-test-id': 'expand-button'
		});
		renderer.act(() => {
			expandButton.props.onClick();
		});
		const referenceInput = testInstance.findAllByProps({
			'data-test-id': 'inline-inputs'
		})[0];
		renderer.act(() => {
			referenceInput.parent.props.onChange('Cenas');
		});
		expect(testProps.reference).toBe('Cenas');
	});
});
