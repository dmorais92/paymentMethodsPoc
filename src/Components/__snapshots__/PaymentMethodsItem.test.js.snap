// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaymentMethodItem component matches snapshot 1`] = `
<li
  className="sc-EHOje jJKfMq"
>
  <div
    className="sc-bZQynM ftWVEo"
  >
    <span
      className="sc-gzVnrw dwvxdF"
    >
      0
    </span>
    <span
      className="sc-htoDjs itHVYD"
    >
      
    </span>
  </div>
  <div
    className="sc-bZQynM ftWVEo"
  >
    <span
      className="sc-gzVnrw dwvxdF"
      data-test-id="reference-title"
    >
      Empty reference
    </span>
    <span
      className="sc-htoDjs itHVYD"
    >
      Today
    </span>
  </div>
  <div
    className="sc-bZQynM ftWVEo"
  >
    <button
      className="sc-dnqmqq dmmmCr"
      onClick={[Function]}
    >
      <svg
        className="sc-gZMcBi gxeeSH"
      >
        DeleteIcon.svg
      </svg>
    </button>
    <button
      className="sc-dnqmqq fSRbzf"
      data-test-id="expand-button"
      onClick={[Function]}
    >
      <svg
        className=""
      >
        DownIcon.svg
      </svg>
    </button>
  </div>
</li>
`;
